src/db.c

```diff
@@ -1070,6 +1070,104 @@ void moveCommand(client *c) {
     addReply(c,shared.cone);
 }
 
+void copyCommand(client *c) {
+    robj *o;
+    redisDb *src, *dst;
+    int srcid;
+    long long dbid, expire;
+    int j, replace = 0, delete = 0;
+    
+    if (c->argc >= 6){
+        addReply(c,shared.syntaxerr);
+        return;
+    }
+    
+    /* Obtain source and target DB pointers 
+     * Default target DB is the same as the source DB 
+     * Parse the REPLACE option and targetDB option. */
+    src = c->db;
+    srcid = c->db->id;
+    dbid = c->db->id;
+    for (j = 3; j < c->argc; j++) {
+        if (!strcasecmp(c->argv[j]->ptr,"replace")) {
+            replace = 1;
+        } else if (getLongLongFromObject(c->argv[j],&dbid) == C_ERR ||
+            dbid < INT_MIN || dbid > INT_MAX ||
+            selectDb(c,dbid) == C_ERR)
+        {
+            addReply(c,shared.outofrangeerr);
+            return;
+        }
+    }
+    dst = c->db;
+    selectDb(c,srcid); /* Back to the source DB */
+    if ((server.cluster_enabled == 1) && (srcid != 0 || dbid != 0)) {
+        addReplyError(c,"Copying to another database is not allowed in cluster mode");
+        return;
+    }
+
+    /* If the user select the same DB as
+     * the source DB and using newkey as the same key
+     * it is probably an error. */
+    robj *key = c->argv[1];
+    robj *newkey = c->argv[2];
+    if (src == dst && (sdscmp(key->ptr, newkey->ptr) == 0)) {
+        addReply(c,shared.sameobjecterr);
+        return;
+    }
+
+    /* Check if the element exists and get a reference */
+    o = lookupKeyWrite(c->db,key);
```

Comment:
lookupKeyRead

Reply:
actually, for some unclear reason, redis uses lookupKeyWrite when reading keys on behalf of a write operation. looking at sunionstore, this intent is clear.
see https://github.com/redis/redis/issues/7475

Reply:
Ok, point taken, it's weird behavior but good to know.